source/dofs/dof_handler.cc

```diff
@@ -1104,7 +1104,7 @@ DoFHandler<dim,spacedim>::renumber_dofs (const std::vector<types::global_dof_ind
 {
   Assert(levels.size()>0, ExcMessage("You need to distribute DoFs before you can renumber them."));
 
-  AssertDimension (new_numbers.size(), n_locally_owned_dofs());
+  AssertDimension (new_numbers.size(), n_dofs());
```

Comment:
This change cannot be correct. I'm pretty sure that the function documents that it wants a vector of length `n_locally_owned_dofs()`, which is the only reasonable thing in the parallel distributed context where we may have billions of unknowns and we could not possibly create a vector of that length.

So the function must expect a vector of that length in the parallel distributed case.

The `parallel::shared::Triangulation` is a corner case because there every processor has knowledge about all DoFs. So one could in principle think of doing something like this:
* If p::d::Triangulation, then require `n_locally_owned_dofs()`
* If p::s::Triangulation, then we could debate which one is correct
* If sequential Triangulation, then the two are the same anyway.

To decide what to do in the second case, your testcase seems to suggest that calling `policy->renumber_dofs()` works with a vector of size `n_dofs()` -- is this correct?

Reply:
On second thought, since `p::s::Triangulation` originated in a scheme where the DoFHandler knows everything and only the matrix and vectors are distributed, I think that it makes sense to expect that the function takes a vector of size `n_dofs()`. So I think what is necessary is to make the assertion check different things depending on what kind of triangulation is being used. You can use that with a `dynamic_cast`, for which I am pretty sure there are examples in this file.